src/games/strategy/engine/framework/ui/NewGameChooser.java

```diff
@@ -238,20 +245,29 @@ public synchronized static NewGameChooserModel getNewGameChooserModel() {
     return s_cachedGameModel;
   }
 
-  // any methods touching s_cachedGameModel should be both static and synchronized
+  /** Starts updating the NewGameChooserModel cache, this can be a lengthy process (seconds) */
   public synchronized static void refreshNewGameChooserModel() {
-    clearNewGameChooserModel();
-    s_cachedGameModel = new NewGameChooserModel();
-  }
+    mapParsingInProgress = true;
+    cacheClearedMessenger = new ClearGameChooserCacheMessenger();
 
-  // any methods touching s_cachedGameModel should be both static and synchronized
-  public synchronized static void clearNewGameChooserModel() {
-    if (s_cachedGameModel != null) {
-      s_cachedGameModel.clear();
-      s_cachedGameModel = null;
+    NewGameChooserModel model = new NewGameChooserModel(cacheClearedMessenger);
+      // Cache might not be populated if another thread has cleared
+      // the cache while we were loading.
+    if (mapParsingInProgress) {
```

Comment:
A lock at lines 255 and 264 around "mapParsingInProgress" should fix that. The lock is being held for just in-memory operations, so it'll be released quickly and won't delay game launch.

There are some scenarios that are impossible since both threads would have to be driven by simultaneous user interaction. I can't recall which scenarios exactly that is the case, but doesn't seem to matter for this specific case.